pkg/ingress/ingress.go

```diff
@@ -0,0 +1,178 @@
+func NewIngressClient(client client.Client, reader client.Reader) *Client {
+	return &Client{
+		client:  client,
+		rClient: reader,
+	}
+}
+
+func (c *Client) fromExtToNet(ingress extv1beta.Ingress) netv1beta.Ingress {
+	oldIngress := netv1beta.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Ingress",
+			APIVersion: "extensions/v1beta1",
+		},
+		ObjectMeta: ingress.ObjectMeta,
+		Spec: netv1beta.IngressSpec{
+			Backend: &netv1beta.IngressBackend{
+				ServiceName: ingress.Spec.Backend.ServiceName,
+				ServicePort: ingress.Spec.Backend.ServicePort,
+			},
+		},
+	}
+
+	for _, tls := range ingress.Spec.TLS {
+		oldIngress.Spec.TLS = append(oldIngress.Spec.TLS, netv1beta.IngressTLS{
+			Hosts:      tls.Hosts,
+			SecretName: tls.SecretName,
+		})
+	}
+
+	for _, rule := range ingress.Spec.Rules {
+		httpIngressPaths := make([]netv1beta.HTTPIngressPath, len(rule.HTTP.Paths))
+		for i, path := range rule.HTTP.Paths {
+			httpIngressPaths[i].Backend.ServicePort = path.Backend.ServicePort
+			httpIngressPaths[i].Backend.ServiceName = path.Backend.ServiceName
+			httpIngressPaths[i].Path = path.Path
+
+		}
+
+		oldIngress.Spec.Rules = append(oldIngress.Spec.Rules, netv1beta.IngressRule{
+			Host: rule.Host,
+			IngressRuleValue: netv1beta.IngressRuleValue{
+				HTTP: &netv1beta.HTTPIngressRuleValue{
+					Paths: httpIngressPaths,
+				},
+			},
+		})
+	}
+
+	return oldIngress
+}
+
+func (c *Client) fromNetToExt(ingress netv1beta.Ingress) extv1beta.Ingress {
+	oldIngress := extv1beta.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Ingress",
+			APIVersion: "extensions/v1beta1",
+		},
+		ObjectMeta: ingress.ObjectMeta,
+	}
+
+	if ingress.Spec.Backend != nil {
+		oldIngress.Spec = extv1beta.IngressSpec{
+			Backend: &extv1beta.IngressBackend{
+				ServiceName: ingress.Spec.Backend.ServiceName,
+				ServicePort: ingress.Spec.Backend.ServicePort,
+			},
+		}
+	}
+
+	for _, tls := range ingress.Spec.TLS {
+		oldIngress.Spec.TLS = append(oldIngress.Spec.TLS, extv1beta.IngressTLS{
+			Hosts:      tls.Hosts,
+			SecretName: tls.SecretName,
+		})
+	}
+
+	for _, rule := range ingress.Spec.Rules {
+		httpIngressPaths := make([]extv1beta.HTTPIngressPath, len(rule.HTTP.Paths))
+		for i, path := range rule.HTTP.Paths {
+			httpIngressPaths[i].Backend.ServicePort = path.Backend.ServicePort
+			httpIngressPaths[i].Backend.ServiceName = path.Backend.ServiceName
+			httpIngressPaths[i].Path = path.Path
+
+		}
+
+		oldIngress.Spec.Rules = append(oldIngress.Spec.Rules, extv1beta.IngressRule{
+			Host: rule.Host,
+			IngressRuleValue: extv1beta.IngressRuleValue{
+				HTTP: &extv1beta.HTTPIngressRuleValue{
+					Paths: httpIngressPaths,
+				},
+			},
+		})
+	}
+
+	return oldIngress
+}
+
+// List is a wrap function that calls k8s client List with extend or networking API.
+func (c *Client) List(ctx context.Context, list *netv1beta.IngressList, opts ...client.ListOption) error {
+	if viper.Get("ingress-api") == ExtensionAPI {
```

Comment:
This could be at client init time. I guess nobody will enable extensions API dynamically